rust/lance-core/src/utils/mask.rs

```diff
@@ -0,0 +1,307 @@
+        if let Some(existing) = self.block_list {
+            Self {
+                block_list: Some(existing | block_list),
+                allow_list: self.allow_list,
+            }
+        } else {
+            Self {
+                block_list: Some(block_list),
+                allow_list: self.allow_list,
+            }
+        }
+    }
+
+    /// Also allow the given ids
+    pub fn also_allow(self, allow_list: RoaringTreemap) -> Self {
+        if let Some(existing) = self.allow_list {
+            Self {
+                block_list: self.block_list,
+                allow_list: Some(existing | allow_list),
+            }
+        } else {
+            Self {
+                block_list: self.block_list,
+                // allow_list = None means "all rows allowed" and so allowing
+                //              more rows is meaningless
+                allow_list: None,
+            }
+        }
+    }
+
+    /// Converst a mask into an arrow array
+    ///
+    /// A row id mask is not very arrow-compatible.  We can't make it a batch with
+    /// two columns because the block list and allow list will have different lengths.  Also,
+    /// there is no Arrow type for compressed bitmaps.
+    ///
+    /// However, we need to shove it into some kind of Arrow container to pass it along the
+    /// datafusion stream.  Perhaps, in the future, we can add row id masks as first class
+    /// types in datafusion, and this can be passed along as a mask / selection vector.
+    ///
+    /// We serialize this as a variable length binary array with two items.  The first item
+    /// is the block list and the second item is the allow list.
+    pub fn into_arrow(&self) -> Result<Arc<dyn Array>> {
+        let block_list_length = self
+            .block_list
+            .as_ref()
+            .map(|bl| bl.serialized_size())
+            .unwrap_or(0);
+        let allow_list_length = self
+            .allow_list
+            .as_ref()
+            .map(|al| al.serialized_size())
+            .unwrap_or(0);
+        let lengths = vec![block_list_length, allow_list_length];
+        let offsets = OffsetBuffer::from_lengths(lengths);
+        let mut value_bytes = vec![0; block_list_length + allow_list_length];
+        let mut validity = vec![false, false];
+        if let Some(block_list) = &self.block_list {
+            validity[0] = true;
+            block_list.serialize_into(&mut value_bytes[0..])?;
+        }
+        if let Some(allow_list) = &self.allow_list {
+            validity[1] = true;
+            allow_list.serialize_into(&mut value_bytes[block_list_length..])?;
+        }
+        let values = Buffer::from(value_bytes);
+        let nulls = NullBuffer::from(validity);
+        Ok(Arc::new(BinaryArray::try_new(
+            offsets,
+            values,
+            Some(nulls),
+        )?))
+    }
+
+    /// Deserialize a row id mask from Arrow
+    pub fn from_arrow(array: &dyn Array) -> Result<Self> {
```

Comment:
is this expect a Int32Array? Should we just use `from(arrow: &Int32Array)` to check the type at compile time.

Reply:
That will just push the check up one level.  The only thing we can pass between DF nodes is a record batch and we will lose the typing at some point.

Still, that might make this interface easier in case we use it from somewhere else.  I will make the change.

Reply:
I've changed to `BinaryArray` (it's not an `Int32Array` because its the serialized compressed bitmaps and not the materialized row ids)